src/utils/agent_tools.py

```
@@ -0,0 +1,1710 @@
+            logger.warning(error_msg)
+            return error_msg
+        except KeyError as e:
+            # Missing required parameters - return to LLM
+            error_msg = f"Tool {tool_name} missing required parameter: {e}"
+            logger.warning(error_msg)
+            return error_msg
+        except Exception as e:
+            # Unexpected errors - log with full traceback but still return to LLM
+            # We don't re-raise because the LLM should be able to continue with other tools
+            error_msg = f"Tool {tool_name} failed unexpectedly: {type(e).__name__}: {e}"
+            logger.error(error_msg, exc_info=True)
```

Ruff (0.14.10)

Logging `.exception(...)` should be used instead of `.error(..., exc_info=True)`

(G201)